common/djangoapps/student/models.py

```diff
@@ -3116,6 +3122,135 @@ class AccountRecoveryConfiguration(ConfigurationModel):
+    to add multiple celebrations for longer streaks.
+    5. We keep track of the longest_ever_streak field for potential future use for badging purposes.
+
+    .. no_pii:
+    """
+    user = models.OneToOneField(User, models.CASCADE, related_name='celebration')
+    first_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    last_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    last_streak_celebration = models.DateField(default=None, null=True, blank=True)
+    longest_ever_streak = models.IntegerField(default=0)
+
+    def __str__(self):
+        return (
+            "[CourseEnrollmentCelebration] user: {}; first_day_of_streak {};"
+            " last_day_of_streak {}; last_streak_celebration {}; longest_ever_streak {};"
+        ).format(self.user.username, self.first_day_of_streak,
+                 self.last_day_of_streak, self.last_streak_celebration, self.longest_ever_streak)
+
+    @classmethod
+    def _get_now(cls):
+        """ Retrieve the value for the current datetime in the user's timezone """
+        # importing here to avoid a circular import
+        from lms.djangoapps.courseware.context_processor import user_timezone_locale_prefs
+        user_timezone_locale = user_timezone_locale_prefs(crum.get_current_request())
+        user_timezone = timezone(user_timezone_locale['user_timezone'] or str(UTC))
+        return user_timezone.localize(datetime.now())
+
+    @classmethod
+    def _calculate_streak_updates(cls, celebration, now):
+        """ Calculate the updates that should me applied to the streak fields of the provided celebration """
+        first_day_of_streak = celebration.first_day_of_streak
+        last_day_of_streak = celebration.last_day_of_streak
+        last_streak_celebration = celebration.last_streak_celebration
+        today = now.date()
+        yesterday = today - timedelta(days=1)
+        should_celebrate = False
+
+        first_ever_streak = None is last_streak_celebration is first_day_of_streak is last_day_of_streak
+        in_active_streak = last_day_of_streak and first_day_of_streak
+        broken_streak = last_day_of_streak not in (yesterday, today)
+        can_start_new_streak = last_day_of_streak and last_day_of_streak + timedelta(days=STREAK_BREAK_LENGTH) < today
+
+        if first_ever_streak:
+            # Start new streak
+            first_day_of_streak = today
+            last_day_of_streak = today
+        elif broken_streak:
+            if can_start_new_streak:
+                # If we were on a break and the break is over start new streak
+                first_day_of_streak = today
+                last_day_of_streak = today
+        elif in_active_streak:
+            # Increment existing streak
+            last_day_of_streak = today
+
+            streak_length = (last_day_of_streak - first_day_of_streak).days + 1
+            if streak_length == STREAK_LENGTH_TO_CELEBRATE:
```

Comment:
This chunk of code could be brought out of this `in_active_streak` block, yeah? If we ever had a 1-day-streak, this would need to be. But fine to assume we wouldn't do that.